### ISSUE TITLE
Resolution bias in fits where bin size is too large (>1/4 of resolution)

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#162]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/162) by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

I attached an image of one example of this bias from some EBIT data, where I'm fitting essentially an infinitely narrow line, so totally dominated by sensors resolution. I can think of two solutions to avoid this bias:

1. throw an error/warning when the fit resolution is too close to the bin size
2. integrate over bins when fitting
3. if 2 is too slow, maybe only integrate over bins if binsize is too big?

![fwhm vs binsize.png](https://bitbucket.org/repo/KrBpjo/images/192517358-fwhm%20vs%20binsize.png)

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Attachment isn't showing.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

I just edited the post and reuplaoded the image. Appears for me now.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Got it now, too. I see the point.

As we discussed on slack, my first reaction is that this is user error. But of course, it's our job to make user error harder to happen and easier to recover from.

Proposal that we liked was to accept a keyword argument to the method `fit(..., integrateNpoints=1)`. If larger than 1, we'll oversample the model with that many points per bin and use the appropriate numerical integration weighting. The obvious choice is 3 points (bin center and 2 edges), which becomes Simpson's rule. I believe it's not hard to generalize to other numbers, at least not to 5 points.

Note to self: Because of the low-E tail and the numerical convolution it requires, we will still need our sub-bin samples to be evenly spaced. (Gaussian quadratures have unevenly spaced samples, and they use that flexibility to achieve more accuracy per sample computed than evenly spaced alternatives.)

We should consider whether it makes sense to add this now or to make it be a part of the planned fitting re-write.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

I think we should put this off to the future.... if the planned fitting re-write happens that would be a good dime. but we should at least let it sit for a few weeks to think about it. I think a combination of you integration points plan and another argument like`min_bins_per_fwhm=15` that throws an error if `resolution/binsize>min_bins_per_fwhm` would force people to notice a problem an explicitly do something about it.

**Comment by joefowler** [comment]
**Original comment by Dan Becker (Bitbucket: [danbek](https://bitbucket.org/danbek), GitHub: [danbek](https://github.com/danbek)).**

----------------------------------------

I'm a little confused about whether you guys are talking about modifying `MaximumLikelihoodHistogramFitter` or `GaussianFitter`. Are these the possible solutions?

 * add a warning to GaussianFitter when bin widths are "too wide" (I don't like this - if the routine is smart enough to recognize that this is a problem, it ought to be smart enough to fix it, especially since we do know how to fix it)
* add a parameter `integrateNpoints` to `GaussianFitter` with a default value of 1 (I don't like this, because if we know that this can be a problem, it should solve the problem by default)
* add a parameter `integrateNpoints` to `MaximumLikelihoodHistogramFitter` with a default value of 1 (I like this because it makes this logic easy to use by all callers, and the default of 1 is correct if some pedantic programmer decides to write his own code calling `MaximumLikelihoodHistogramFitter` that does the exact integral). In this case I would suggest modifying Gaussian fitter to use this parameter by default.

Cogitating on this for a few weeks sounds fine to me, as does postponing to a fitting re-write.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

I take Dan's point about not wanting to provide mere warnings or optional arguments that will solve the bias, when we already know how to detect and fix the bias.

However, Galen is arguing that the bias arises only when the ratio of bin size to resolution is too large. And unfortunately, when you start the fit, you do not know the resolution yet! (Not unless you have frozen that fit parameter for this specific fit.) That will make it hard to automate the choice of integrate vs sample.

Also, and as always, we have to figure out how to balance the desire for fits to Just Work against the desire for fits to be fully controllable when we need it. (Maybe the caller would happily accept a small bias on the resolution if the fit would be 3x faster.)

Proposal:

* `LineFitter.fit` method takes a new, optional argument `integrate_n_points=None`.
* If the argument is a positive integer, use that many points per bin (1=midpoint rule; 2=trapezoid rule; 3, 5, 7, ...=Simpson's rule over each 3-point set; 4, 6, 8... = ???). I would have to research whether trapezoid actually helps, and what to do with the larger even numbers. Perhaps we only allow odd number arguments? Or add 1 to any even number argument?
* If argument is explicitly given as 1 or 2 and the bin size / resolution ratio comes out to be bigger than ~0.1, then issue a warning.
* If argument is the default value of `None`, then use a heuristic like this: use the initial guess at resolution to estimate bin size / res. If less than 0.1, use 1 points; if bigger than 0.5, use 5 (or more??) points; if between, use 3.
* When using a rule other than the midpoint rule, wrap the model function with one that takes the model shape and performs Simpson's rule, or whatever, so that the wrapped function can work as if we'd changed nothing.
* I think that in this proposal, the underlying `MaximumLikelihoodHistogramFitter` class doesn't need to change.

The heuristic can fail, say if the initial resolution guess is 11 eV with 1 eV bins, but the final fit value comes out to be 4 eV resolution. We start out believe the ratio is 0.09, but it turns out to be 0.25. The first makes 1-point integration seem okay, but the reality will say otherwise. What to do about that?

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

To repeat some of Galen's plot, above, I simulated some Gaussian fits. Specifically, I ran `several_sims()` as seen below:

```python
import mass
import numpy as np
import pylab as plt

def sim(fwhm=4, N=10000, nsim=10):
    sigma = fwhm/2.3548
    fitter = mass.GaussianFitter()

    p = np.zeros(nsim, dtype=float)
    for i in range(nsim):
        x = np.random.standard_normal(N)*sigma
        c, b = np.histogram(x, 100, [-50, 50])
        param, covar = fitter.fit(c, b, [fwhm, 0, c.max(), 0, 0, 0, 25],
                        plot=False, hold=(3, 4, 5, 6))
        p[i] = param[0]
    plt.clf()
    plt.hist(p, 100)
    print np.median(p), p.mean(), mass.robust.trimean(p)
    return mass.robust.trimean(p)


def several_sims(nsim=100):
    widths = np.array([.7, .8, 1, 1.2, 1.5, 2, 2.5, 3, 3.5, 4, 5, 6, 8, 10, 12, 15, 20, 30, 40])
    fitwidth = 0*widths
    for i, fwhm in enumerate(widths):
        fitwidth[i] = sim(fwhm, 10000, nsim=nsim)

    plt.clf()
    plt.semilogx(widths, fitwidth/widths, "-or")
    plt.xlabel("Gaussian FWHM/bin width ratio")
    plt.ylabel("Ratio: Fit-result FWHM / true FWHM")
    plt.plot([widths[0], widths[-1]], [1, 1], color='gray')
    plt.title("Gaussian width overestimate vs (width/binsize) ratio")
```

The plot shows the ratio by which we overestimate the Gaussian FWHM of a sim with zero background counts and 10,000 zero-mean Gaussian-distributed data, as a function of the true FWHM. In all cases, the fitted data contain 100 bins. The bins are 1 unit wide and run from [-50, 50] units. This is kind of Galen's plot in reverse, because I varied the Gaussian width and fixed the bin size, while he did the opposite.

Specifically, I ran a `GaussianFitter()` with only the three parameters {FWHM, center, amplitude} allowed to vary (no tail, no background). I think Galen's left-most (most biased) point corresponds to (2, 1.05) on this plot: a 5% effect when the Gaussian is only 2x wider than each bin.

I'll be curious to see what we can make this plot look like with numerical integration.

![gaussian_bias.png](https://bitbucket.org/repo/KrBpjo/images/2399101608-gaussian_bias.png)

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Ooh, I just realized one possible problem. In the limit that we have FWHM << bin width, so that only one bin contains any non-background counts, there is a complete degeneracy between width and amplitude, even if we are doing arbitrarily dense, arbitrarily perfect numerical integration of the theory model curve. Hmm.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Perform numerical integration in fits, if needed

Fixes #162. Uses a heuristic to decide how much to integrate over bins by default, or accepts user's integrate_n_points argument. Tests are also included.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Removing milestone: v0.7 (automated comment)
